### DISCUSSION TITLE
Proposal : Operator overloading change from static method to instance method

### DISCUSSION THREAD

**Discussion by NutzNK** [dotnet/csharplang#3218] [Language Ideas]
Now Operator overloading has pattern like this.

```C#
public static Int32 operator +(Int32 left, Int32 right) 
{
   return new Int32(){ value = left.value + right.value };
}
```

I recommend to change it to instance method like this

```C#
public Int32 operator +(Int32 other) 
{
   //if(this is null || other is null) throw; //this code for RefType checking only
   return new Int32(){ value = this.value + other.value };
}
```

Because all operation overloading have first argument that act like instance method.

Benefit as I can think now:
1. Interface can assign operator.
2. Generics type can implement arithmetic and comparison operator because of benefit from 1st item.
3. Operator overloading can apply to extension method.

**Comment by Unknown6656** [comment]
I do not like this syntax (I did not like it in C++, either), because one does not always know which instance is which (meaning, the left-sided or right-sided operand).

I would rather see static methods/operators being supported by interfaces. This should be covered by the proposal #164.

**Comment by HaloFour** [comment]
#2065

https://blogs.msdn.microsoft.com/ericlippert/2007/05/14/why-are-overloaded-operators-always-static-in-c/

**Comment by NutzNK** [comment]
@HaloFour Thank you for your information, I under stand what is their concern.
But now programming trending is numerical computing like AI, ML, other data analytics.
But C# still cannot make good productivity for numerical coding like
```C#
var vec = Vector<double>(array);
var result = vector * 2.0;
```
That why so many people prefer to develop numerical computing on python.

There are some other patterns can fix this issue too.
operator overloading on extension method can solve this problem but this way need to change a lot in this language too.

**Comment by HaloFour** [comment]
@NutzNK 

https://github.com/dotnet/csharplang/issues/2065#issuecomment-446725672

**Comment by TonyValenti** [comment]
I would much rather see the ability for overloads to be extension methods.
